src/main/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/DeterministicShuffleShardSyncLeaderDecider.java

```diff
@@ -0,0 +1,184 @@
+     * @param leaseManager LeaseManager instance.
+     */
+    DeterministicShuffleShardSyncLeaderDecider(KinesisClientLibConfiguration config,
+            ILeaseManager<KinesisClientLease> leaseManager) {
+        this(config, leaseManager, PERIODIC_SHARD_SYNC_MAX_WORKERS_DEFAULT);
+    }
+
+    /**
+     * Create an instance overriding the default periodic shard sync worker count.
+     * This is public for use outside KCL.
+     *
+     * @param config                      KinesisClientLibConfiguration instance
+     * @param leaseManager                LeaseManager instance.
+     * @param numPeriodicShardSyncWorkers Max number of periodic shard sync workers.
+     */
+    public DeterministicShuffleShardSyncLeaderDecider(KinesisClientLibConfiguration config,
+            ILeaseManager<KinesisClientLease> leaseManager, int numPeriodicShardSyncWorkers) {
+        this(config, leaseManager, Executors.newScheduledThreadPool(LEADER_DECIDER_THREAD_COUNT),
+                numPeriodicShardSyncWorkers);
+    }
+
+    DeterministicShuffleShardSyncLeaderDecider(KinesisClientLibConfiguration config,
+            ILeaseManager<KinesisClientLease> leaseManager, ScheduledExecutorService leaderElectionThreadPool,
+            int numPeriodicShardSyncWorkers) {
+        this.config = config;
+        this.leaseManager = leaseManager;
+        this.leaderElectionThreadPool = leaderElectionThreadPool;
+        this.numPeriodicShardSyncWorkers = numPeriodicShardSyncWorkers;
+    }
+
+    /*
+     * Shuffles the leases deterministically and elects numPeriodicShardSyncWorkers number of workers
+     * as leaders (workers that will perform shard sync).
+     */
+    private void electLeaders() {
+        try {
+            LOG.debug("Started leader election: " + System.currentTimeMillis());
+            List<KinesisClientLease> leases = leaseManager.listLeases();
+            List<String> uniqueHosts = leases.stream().map(KinesisClientLease::getLeaseOwner)
+                    .filter(owner -> owner != null).distinct().sorted().collect(Collectors.toList());
+
+            Collections.shuffle(uniqueHosts, new Random(DETERMINISTIC_SHUFFLE_SEED));
+            int numShardSyncWorkers = Math.min(uniqueHosts.size(), numPeriodicShardSyncWorkers);
+            // In case value is currently being read, we wait for reading to complete before updating the variable.
+            // This is to prevent any ConcurrentModificationException exceptions.
+            readWriteLock.writeLock().lock();
+            leaders = new HashSet<>(uniqueHosts.subList(0, numShardSyncWorkers));
+            LOG.debug("Elected leaders: " + String.join(", ", leaders));
+            LOG.debug("Completed leader election: " + System.currentTimeMillis());
+        } catch (DependencyException | InvalidStateException | ProvisionedThroughputException e) {
+            LOG.error("Exception occurred while trying to fetch all leases for leader election", e);
```

Comment:
What is happening during these exceptions? Is it getting retried? These errors seem like known conditions, and safe thing is to retry, so it can be warning instead of error

Reply:
Retries are within AWS SDK. If it's still thrown, it means retries couldn't get past the exception. Logging an error makes sense in that case.

Reply:
In that case we could probably keep a count of how many times did the retry fail consecutively and if it more that X times then log as errors.

Reply:
As discussed, we are going with the implemented design for now.